XMLSerializer.serializeToString() should support Attr node as parameter

Ahmad-S792 opened the pull request:
#### 38ce5b803b349852d5800b8027e123ca8248e30f
<pre>
XMLSerializer.serializeToString() should support Attr node as parameter

XMLSerializer.serializeToString() should support Attr node as parameter
<a href="https://bugs.webkit.org/show_bug.cgi?id=247096">https://bugs.webkit.org/show_bug.cgi?id=247096</a>

Reviewed by Ryosuke Niwa.

This is to align Webkit with Blink / Chrome.

Inspired from - <a href="https://chromium.googlesource.com/chromium/src.git/+/271184ec08ca181a7d93eed532f3ce9efb9c0bcc">https://chromium.googlesource.com/chromium/src.git/+/271184ec08ca181a7d93eed532f3ce9efb9c0bcc</a>

The specification [1] doesn&apos;t define the behavior for Attr because the
specification assumes Attr is not a Node. However, passing an Attr is possible
now and we should not have an assertion failure.

This patch is aimed to serializeToString() matching to Web Specification.
It means serializing the attribute value.

[1] <a href="https://w3c.github.io/DOM-Parsing/#dfn-concept-serialize-xml">https://w3c.github.io/DOM-Parsing/#dfn-concept-serialize-xml</a>

* Source/WebCore/editing/MarkupAccumulator.cpp: Added &quot;Attr.h&quot; header
(MarkAccumulator::appendNonElementNode): Add result value for &quot;Attribute Node&quot;
* LayoutTests/fast/dom/XMLSerializer.html: Updated to use &quot;testharness&quot; and Attr test as well
* LayoutTests/fast/dom/XMLSerializer-expected.txt: Added Test Case Expectations

Canonical link: <a href="https://commits.webkit.org/256126@main">https://commits.webkit.org/256126@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2432e72a83e7f2253e6cde8e60eb145d34eaf473

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94572 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27465 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104210 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164482 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98567 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3797 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31933 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86889 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100179 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100239 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2732 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80950 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29756 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84657 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/84221 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72651 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38335 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18042 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36196 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19316 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4228 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41962 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42058 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38545 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -599,7 +600,9 @@ void MarkupAccumulator::appendNonElementNode(StringBuilder& result, const Node&
         break;
     case Node::ATTRIBUTE_NODE:
-        ASSERT_NOT_REACHED();
+        // Only XMLSerializer can pass an Attr. So, |documentIsHTML| flag is
```
Looks like this can fit in single line.

Ahmad-S792 replied:
Done

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2004-2017 Apple Inc. All rights reserved.
+ * Copyright (C) 2004-2022 Apple Inc. All rights reserved.
  * Copyright (C) 2009, 2010 Google Inc. All rights reserved.
```
Given this change is inspired by a change in chromium made by a Googler, we should probably attribute the copyright to Google as well.

Ahmad-S792 replied:
Done - updated Google to 2009-2022 and reverted Apple back to original.

rniwa reviewed (approved):
r=me

Ahmad-S792 committed:
XMLSerializer.serializeToString() should support Attr node as parameter

XMLSerializer.serializeToString() should support Attr node as parameter
https://bugs.webkit.org/show_bug.cgi?id=247096

Reviewed by Ryosuke Niwa.

This is to align Webkit with Blink / Chrome.

Inspired from - https://chromium.googlesource.com/chromium/src.git/+/271184ec08ca181a7d93eed532f3ce9efb9c0bcc

The specification [1] doesn't define the behavior for Attr because the
specification assumes Attr is not a Node. However, passing an Attr is possible
now and we should not have an assertion failure.

This patch is aimed to serializeToString() matching to Web Specification.
It means serializing the attribute value.

[1] https://w3c.github.io/DOM-Parsing/#dfn-concept-serialize-xml

* Source/WebCore/editing/MarkupAccumulator.cpp: Added "Attr.h" header
(MarkAccumulator::appendNonElementNode): Add result value for "Attribute Node"
* LayoutTests/fast/dom/XMLSerializer.html: Updated to use "testharness" and Attr test as well
* LayoutTests/fast/dom/XMLSerializer-expected.txt: Added Test Case Expectations

Canonical link: https://commits.webkit.org/256126@main
